SCClassLibrary/Common/Core/CondVar.sc

```diff
@@ -0,0 +1,105 @@
+CondVar {
+	var waitingThreads, timeoutThreads;
+
+	*new { ^super.newCopyArgs(Array.new, Array.new) }
+
+	wait { |predicate|
+		if(predicate.notNil) {
+			while { predicate.value.not } { this.wait }
+		} {
+			this.prSleepThread(thisThread);
+			\wait.yield
+		}
+	}
+
+	waitFor { |timeoutBeats, predicate|
+		var endingBeats, waitingThread, timeoutThread, enterBeats = thisThread.beats;
+		var didNotTimeout = true;
+
+		// May throw if preconditions are not met
+		timeoutBeats = this.prHandleTimeoutPreconditions(timeoutBeats);
+
+		if(timeoutBeats <= 0) { ^predicate.notNil and: { predicate.value } };
+
+		if(predicate.notNil) {
+			endingBeats = enterBeats + timeoutBeats;
+			// will exit when predicate is true or timeout has expired
+			while {
+				predicate.value.not
+			} {
+				if(this.waitFor(endingBeats - thisThread.beats).not) { ^predicate.value }
+			};
+
+			^true // we were signalled and predicate was true
+		} {
+			waitingThread = thisThread;
+			timeoutThread = Routine {
+				var wokenThread;
+				// TODO is this right? can we be more precise?
+				timeoutBeats.wait;
+
+				// wokenThread must be non-nil. If the thread woke before this timeout fired, this routine will already
+				// be stopped and we won't get here.
+				didNotTimeout = false;
+				wokenThread = this.prRemoveWaitingThread(waitingThread);
+				timeoutThreads.remove(timeoutThread);
+				this.prWakeThread(wokenThread);
+			};
+			timeoutThreads = timeoutThreads.add(timeoutThread);
+			timeoutThread.play(thisThread.clock);
+			this.wait;
+			// TODO will this be gc'd properly?
+			timeoutThreads.remove(timeoutThread).stop;
+
+			^didNotTimeout
+		}
+	}
+
+	signalOne {
```

Comment:
in similar cases in sclang, we have pairings of `something` and `somethingAll` – so just `signal` might be more idiomatic.

Reply:
I don't mind doing this. The old interface used `unhang` and `notify` to mean "unhang all" and "notify all", so I thought it would be better to be explicit here in case people are used to the old interface.

Reply:
I see, that makes sense. Still the word "one" immediately rises the question for me: "which one?" Should it be `signalFirst` or `signalNext`?

Reply:
There are a lot of condition variable implementations out there we can draw from for inspiration. I'd encourage you to go take a look at some and then see if you still feel the same way about these names. The ones I looked at while reviewing James's PR were the implementations from C++ (which is the one closest to the design I used here), WinAPI, pthreads, Java, D, Rust, Python, and Go.

Reply:
Ok, good! I've checked some, and I see that notify_one seems to be a model for c++ oriented implementations. In Ruby and Java, the combination seems to be  `signal` and `broadcast`, which sounds good. But I see your point, so why not keep it.